extensions/parquet/base/src/main/java/io/deephaven/parquet/base/ColumnPageReaderImpl.java

```diff
@@ -110,53 +122,66 @@ public int readRowCount() throws IOException {
     @Override
     public IntBuffer readKeyValues(IntBuffer keyDest, int nullPlaceholder) throws IOException {
         try (final SeekableByteChannel readChannel = channelsProvider.getReadChannel(filePath)) {
-            readChannel.position(offset);
             ensurePageHeader(readChannel);
             return readKeyFromDataPage(keyDest, nullPlaceholder, readChannel);
         }
     }
 
     /**
-     * If {@link #pageHeader} is {@code null}, read it from the file and increment the {@link #offset} and file position
-     * by the length of page header. This method assumes that file position is set to {@link #offset} before calling.
-     * This method also read the number of values in the page from the header.
+     * If {@link #pageHeader} is {@code null}, read it from the file, and increment the {@link #offset} by the length of
+     * page header. Channel position would be set to the end of page header or beginning of data before returning.
      */
-    private synchronized void ensurePageHeader(SeekableByteChannel file) throws IOException {
-        if (pageHeader == null) {
-            if (file.position() != offset) {
-                throw new IllegalStateException("File position = " + file.position() + " not equal to expected offset ="
-                        + offset);
-            }
-            int maxHeader = START_HEADER;
-            boolean success;
-            do {
-                ByteBuffer headerBuffer = ByteBuffer.allocate(maxHeader);
-                file.read(headerBuffer);
-                headerBuffer.flip();
-                ByteBufferInputStream bufferedIS = ByteBufferInputStream.wrap(headerBuffer);
-                try {
-                    pageHeader = Util.readPageHeader(bufferedIS);
-                    offset += bufferedIS.position();
-                    success = true;
-                } catch (IOException e) {
-                    success = false;
-                    if (maxHeader > MAX_HEADER) {
-                        throw e;
+    private void ensurePageHeader(final SeekableByteChannel file) throws IOException {
+        // Set this channel's position to appropriate offset for reading. If pageHeader is null, this offset would be
+        // the offset of page header, else it would be the offset of data.
+        file.position(offset);
```

Comment:
Should this be inside the synchronized block? I guess the file is never shared, and consequently need not have access synchronized? So I think the answer is "no"?

Reply:
Yes, that's right. And it wasn't inside the synchronized block earlier, so keeping it consistent.